### PR TITLE
Fix #20 - Assign protoList to Array

### LINKED ISSUE #20: Type checker considers a proto list and array different types
````sh
repos: array of CodeRepository = proj.code_repositories;
````

leads to the compile error:

````
test.boa: compilation failed: Encountered typecheck error at line 1, column 33. incorrect type 'protolist of CodeRepository' for assignment to 'repos: array of CodeRepository'
repos: array of CodeRepository = input.code_repositories;
                                 ^^^^^^^^^^^^^^^^^^^^^^^
````

In terms of documentation, we consider a 'protolist' and 'array' the same.  The compiler still considers these separate types.  We need to fix the mismatch.

Another example of where this fails is with passing protolist's to user-defined functions:

````sh
id := function(fields: array of Variable) ...
id(node.fields);
````

### PR DESCRIPTION
**Pull request by CheShianHung**
Fix #20 - Assign protoList to Array

### CODE REVIEW

File: src/java/boa/types/BoaArray.java [hunk 1]
```diff
@@ -59,6 +59,9 @@ public boolean assigns(final BoaType that) {
 			return true;
 		}
 
+		if (that instanceof BoaProtoList)
```
Comment (psybers) [inline src/java/boa/types/BoaArray.java]: This isnt sufficient.  We also want to trap assignment of the form `array of Expression` and assigning to it `protolist of Statement` but this current check allows assigning **any** protolist to **any** array.

File: test/typecheck/protolist-vs-array.boa [hunk 2]
```diff
@@ -0,0 +1,6 @@
+o: output collection of int;
+repos: array of CodeRepository = input.code_repositories;
+id := function(fields: array of CodeRepository){
+  o << 1;
+};
+id(repos);
```
Comment (psybers) [inline test/typecheck/protolist-vs-array.boa]: You should call `id()` with `input.code_repositories` otherwise there is no reason it would fail, even with the old compiler.

File: src/java/boa/compiler/visitors/TypeCheckingVisitor.java [hunk 3]
```diff
@@ -955,6 +955,9 @@ public void visit(final VarDeclStatement n, final SymbolTable env) {
 
 			if (lhs instanceof BoaArray && rhs instanceof BoaTuple)
 				rhs = new BoaArray(((BoaTuple)rhs).getMember(0));
+			if (lhs instanceof BoaArray && rhs instanceof BoaProtoList){
+				((BoaProtoList)rhs).setType(lhs);
```
Comment (psybers) [inline src/java/boa/compiler/visitors/TypeCheckingVisitor.java]: This is just wrong.  The proto list already has a type set (when it was created).  Why is this here?

File: src/java/boa/compiler/visitors/CodeGeneratingVisitor.java [hunk 4]
```diff
@@ -1386,6 +1386,12 @@ public void visit(final VarDeclStatement n) {
 			}
 		}
 
+		if (lhsType instanceof BoaArray && t instanceof BoaProtoList){
+			String[] rhsType = ((BoaProtoList)t).getType().toString().split(" ");
+			System.out.println("rhs type: " + rhsType[2]);
```
Comment (psybers) [inline src/java/boa/compiler/visitors/CodeGeneratingVisitor.java]: remove debugging statements

File: src/java/boa/compiler/visitors/CodeGeneratingVisitor.java [hunk 5]
```diff
@@ -1386,6 +1386,12 @@ public void visit(final VarDeclStatement n) {
...
+			src += ".toArray(new boa.types.Code." + rhsType[2] + "[" + src + ".size()])";
```
Comment (psybers) [inline src/java/boa/compiler/visitors/CodeGeneratingVisitor.java]: Nope.  We dont want to actually convert things to array!  Just pass the list around as is...

**Comment by psybers** [comment]
Your overall strategy seems wrong.  We don't want to convert the protolists into arrays.  We just want the compiler to 'play nice' and allow them to act like arrays.

File: src/java/boa/compiler/visitors/CodeGeneratingVisitor.java (hunk 5 above)
Comment (CheShianHung) [reply src/java/boa/compiler/visitors/CodeGeneratingVisitor.java]: If the list stays as it is, this error will show up in codegen instead:
`./compile/boa/ProtolistVsArray.java:139: error: incompatible types: List<CodeRepository> cannot be converted to CodeRepository[]
				___repos =  _input .getCodeRepositoriesList();`

Comment (psybers) [reply src/java/boa/compiler/visitors/CodeGeneratingVisitor.java]: Right, because the codegen isnt properly handling the variable declaration.  The point of this issue was to let the compiler think protolists are an array, but not actually codegen as arrays.  They are and always will be lists (we cant control that - protocol buffers create these for us).

File: src/java/boa/types/BoaArray.java (hunk 1 above)
Comment (CheShianHung) [reply src/java/boa/types/BoaArray.java]: I there any other assignment that we want to trap? Or just `array of x` to `protolist of x` and `array of Expression` to `protolist of Statement`?
